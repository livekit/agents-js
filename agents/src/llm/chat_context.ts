// SPDX-FileCopyrightText: 2024 LiveKit, Inc.
//
// SPDX-License-Identifier: Apache-2.0
import type { AudioFrame, VideoFrame } from '@livekit/rtc-node';
import type { CallableFunctionResult, FunctionCallInfo } from './function_context.js';

export enum ChatRole {
  SYSTEM,
  USER,
  ASSISTANT,
  TOOL,
}

export interface ChatImage {
  image: string | VideoFrame;
  inferenceWidth?: number;
  inferenceHeight?: number;
  /**
   * @internal
   * Used by LLM implementations to store a processed version of the image for later use.
   */
  cache: { [id: string | number | symbol]: any };
}

export interface ChatAudio {
  frame: AudioFrame | AudioFrame[];
}

export type ChatContent = string | ChatImage | ChatAudio;

const defaultCreateChatMessage = {
  text: '',
  images: [],
  role: ChatRole.SYSTEM,
};

export class ChatMessage {
  readonly role: ChatRole;
  readonly id?: string;
  readonly name?: string;
  readonly content?: ChatContent | ChatContent[];
  readonly toolCalls?: FunctionCallInfo[];
  readonly toolCallId?: string;
  readonly toolException?: Error;

  /** @internal */
  constructor({
    role,
    id,
    name,
    content,
    toolCalls,
    toolCallId,
    toolException,
  }: {
    role: ChatRole;
    id?: string;
    name?: string;
    content?: ChatContent | ChatContent[];
    toolCalls?: FunctionCallInfo[];
    toolCallId?: string;
    toolException?: Error;
  }) {
    this.role = role;
    this.id = id;
    this.name = name;
    this.content = content;
    this.toolCalls = toolCalls;
    this.toolCallId = toolCallId;
    this.toolException = toolException;
  }

  static createToolFromFunctionResult(func: CallableFunctionResult): ChatMessage {
    if (!func.result && !func.error) {
      throw new TypeError('CallableFunctionResult must include result or error');
    }

    return new ChatMessage({
      role: ChatRole.TOOL,
      name: func.name,
      content: func.result || `Error: ${func.error}`,
      toolCallId: func.toolCallId,
      toolException: func.error,
    });
  }

  static createToolCalls(toolCalls: FunctionCallInfo[], text = '') {
    return new ChatMessage({
      role: ChatRole.ASSISTANT,
      toolCalls,
      content: text,
    });
  }

  static create(
    options: Partial<{
      text?: string;
      images: ChatImage[];
      role: ChatRole;
    }>,
  ): ChatMessage {
    const { text, images, role } = { ...defaultCreateChatMessage, ...options };

    if (!images.length) {
      return new ChatMessage({
        role,
        content: text,
      });
    } else {
      return new ChatMessage({
        role,
        content: [...(text ? [text] : []), ...images],
      });
    }
  }

  /** Returns a structured clone of this message. */
  copy(): ChatMessage {
    return new ChatMessage({
      role: this.role,
      id: this.id,
      name: this.name,
      content: this.content,
      toolCalls: this.toolCalls,
      toolCallId: this.toolCallId,
      toolException: this.toolException,
    });
  }
}

export class ChatContext {
  messages: ChatMessage[] = [];
  metadata: { [id: string]: any } = {};

  append(msg: { text?: string; images?: ChatImage[]; role: ChatRole }): ChatContext {
    this.messages.push(ChatMessage.create(msg));
    return this;
  }

  /** Returns a structured clone of this context. */
  copy(): ChatContext {
    const ctx = new ChatContext();
    ctx.messages.push(...this.messages.map((msg) => msg.copy()));
    ctx.metadata = structuredClone(this.metadata);
    return ctx;
  }
}
