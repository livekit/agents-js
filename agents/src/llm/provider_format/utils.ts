// SPDX-FileCopyrightText: 2025 LiveKit, Inc.
//
// SPDX-License-Identifier: Apache-2.0
import { log } from '../../log.js';
import type {
  ChatContext,
  ChatItem,
  ChatMessage,
  FunctionCall,
  FunctionCallOutput,
} from '../chat_context.js';

class ChatItemGroup {
  message?: ChatMessage;
  toolCalls: FunctionCall[];
  toolOutputs: FunctionCallOutput[];
  logger = log();

  constructor(params: {
    message?: ChatMessage;
    toolCalls: FunctionCall[];
    toolOutputs: FunctionCallOutput[];
  }) {
    this.message = params.message;
    this.toolCalls = params.toolCalls;
    this.toolOutputs = params.toolOutputs;
  }

  static create(params?: {
    message?: ChatMessage;
    toolCalls?: FunctionCall[];
    toolOutputs?: FunctionCallOutput[];
  }) {
    const { message, toolCalls = [], toolOutputs = [] } = params ?? {};
    return new ChatItemGroup({ message, toolCalls, toolOutputs });
  }

  get isEmpty() {
    return (
      this.message === undefined && this.toolCalls.length === 0 && this.toolOutputs.length === 0
    );
  }

  add(item: ChatItem) {
    if (item.type === 'message') {
      if (this.message) {
        throw new Error('only one message is allowed in a group');
      }
      this.message = item;
    } else if (item.type === 'function_call') {
      this.toolCalls.push(item);
    } else if (item.type === 'function_call_output') {
      this.toolOutputs.push(item);
    }
    return this;
  }

  removeInvalidToolCalls() {
    if (this.toolCalls.length === this.toolOutputs.length) {
      return;
    }

    const toolCallIds = new Set(this.toolCalls.map((call) => call.callId));
    const toolOutputIds = new Set(this.toolOutputs.map((output) => output.callId));

    // intersection of tool call ids and tool output ids
    const validCallIds = intersection(toolCallIds, toolOutputIds);

    // filter out tool calls that don't have a corresponding tool output
    this.toolCalls = this.toolCalls.filter((call) => {
      if (validCallIds.has(call.callId)) return true;
      this.logger.warn(
        {
          callId: call.callId,
          toolName: call.name,
        },
        'function call missing the corresponding function output, ignoring',
      );
      return false;
    });

    // filter out tool outputs that don't have a corresponding tool call
    this.toolOutputs = this.toolOutputs.filter((output) => {
      if (validCallIds.has(output.callId)) return true;
      this.logger.warn(
        {
          callId: output.callId,
          toolName: output.name,
        },
        'function output missing the corresponding function call, ignoring',
      );
      return false;
    });
  }

  flatten() {
    const items: ChatItem[] = [];
    if (this.message) items.push(this.message);
    items.push(...this.toolCalls, ...this.toolOutputs);
    return items;
  }
}

function intersection<T>(set1: Set<T>, set2: Set<T>): Set<T> {
  return new Set([...set1].filter((item) => set2.has(item)));
}

/**
 * Group chat items (messages, function calls, and function outputs)
 * into coherent groups based on their item IDs and call IDs.
 *
 * Each group will contain:
 * - Zero or one assistant message
 * - Zero or more function/tool calls
 * - The corresponding function/tool outputs matched by call_id
 *
 * User and system messages are placed in their own individual groups.
 *
 * @param chatCtx - The chat context containing all conversation items
 * @returns A list of ChatItemGroup objects representing the grouped conversation
 */
export function groupToolCalls(chatCtx: ChatContext) {
  const itemGroups: Record<string, ChatItemGroup> = {};
  const insertionOrder: Record<string, number> = {};
  const toolOutputs: FunctionCallOutput[] = [];
  const logger = log();

  let insertionIndex = 0;
  for (const item of chatCtx.items) {
    const isAssistantMessage = item.type === 'message' && item.role === 'assistant';
    const isFunctionCall = item.type === 'function_call';
    const isFunctionCallOutput = item.type === 'function_call_output';

    if (isAssistantMessage || isFunctionCall) {
      // only assistant messages and function calls can be grouped
      const groupId = item.id.split('/')[0]!;
      if (itemGroups[groupId] === undefined) {
        itemGroups[groupId] = ChatItemGroup.create();

        // we use insertion order to sort the groups as they are added to the context
        // simulating the OrderedDict in python
        insertionOrder[groupId] = insertionIndex;
        insertionIndex++;
      }
      itemGroups[groupId]!.add(item);
    } else if (isFunctionCallOutput) {
      toolOutputs.push(item);
    } else {
      itemGroups[item.id] = ChatItemGroup.create().add(item);
    }
  }

  // add tool outputs to their corresponding groups
  const callIdToGroup: Record<string, ChatItemGroup> = {};
  for (const group of Object.values(itemGroups)) {
    for (const toolCall of group.toolCalls) {
      callIdToGroup[toolCall.callId] = group;
    }
  }

  for (const toolOutput of toolOutputs) {
    const group = callIdToGroup[toolOutput.callId];
    if (group === undefined) {
      logger.warn(
        { callId: toolOutput.callId, toolName: toolOutput.name },
        'function output missing the corresponding function call, ignoring',
      );
      continue;
    }
    group.add(toolOutput);
  }

  // validate that each group and remove invalid tool calls and tool outputs
  for (const group of Object.values(itemGroups)) {
    group.removeInvalidToolCalls();
  }

  // sort groups by their item id
  const orderedGroups = Object.entries(itemGroups)
    .sort((a, b) => insertionOrder[a[0]]! - insertionOrder[b[0]]!)
    .map(([, group]) => group);
  return orderedGroups;
}
