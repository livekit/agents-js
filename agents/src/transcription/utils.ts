import type {
  LocalParticipant,
  RemoteParticipant,
  Room,
  TrackPublication,
} from '@livekit/rtc-node';
import { TrackSource } from '@livekit/rtc-node';
import { v4 as uuidv4 } from 'uuid';

export function findMicroTrackId(room: Room, identity: string): string {
  let p: RemoteParticipant | LocalParticipant | undefined = room.remoteParticipants.get(identity);

  if (identity === room.localParticipant?.identity) {
    p = room.localParticipant;
  }

  if (!p) {
    throw new Error(`participant ${identity} not found`);
  }

  // find first micro track
  let trackId: string | undefined;
  p.trackPublications.forEach((track: TrackPublication) => {
    if (track.source === TrackSource.SOURCE_MICROPHONE) {
      trackId = track.sid;
      return;
    }
  });

  if (!trackId) {
    throw new Error(`participant ${identity} does not have a microphone track`);
  }

  return trackId;
}

export function segmentUuid(): string {
  return 'SG_' + uuidv4().replace(/-/g, '').substring(0, 12);
}
