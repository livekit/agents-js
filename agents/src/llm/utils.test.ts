// SPDX-FileCopyrightText: 2025 LiveKit, Inc.
//
// SPDX-License-Identifier: Apache-2.0
import { VideoBufferType, VideoFrame } from '@livekit/rtc-node';
import sharp from 'sharp';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { ChatContext, ChatMessage, type ImageContent } from './chat_context.js';
import { computeChatCtxDiff, serializeImage } from './utils.js';

function createChatMessage(
  id: string,
  content: string,
  role: 'user' | 'assistant' | 'system' = 'user',
): ChatMessage {
  return ChatMessage.create({
    id,
    content,
    role,
  });
}

function createChatContext(messages: ChatMessage[]): ChatContext {
  const ctx = new ChatContext();
  for (const message of messages) {
    ctx.items.push(message);
  }
  return ctx;
}

function createImageContent(
  image: string | VideoFrame,
  inferenceDetail: 'auto' | 'high' | 'low' = 'auto',
  options?: {
    mimeType?: string;
    inferenceWidth?: number;
    inferenceHeight?: number;
  },
): ImageContent {
  return {
    id: 'test-id',
    type: 'image_content',
    image,
    inferenceDetail,
    mimeType: options?.mimeType,
    inferenceWidth: options?.inferenceWidth,
    inferenceHeight: options?.inferenceHeight,
    _cache: {},
  };
}

async function decodeImageToRaw(base64Data: string) {
  const imageBuffer = Buffer.from(base64Data, 'base64');
  const decodedImage = await sharp(imageBuffer).raw().toBuffer({ resolveWithObject: true });
  return { imageBuffer, decodedImage };
}

function createSolidColorFrame(
  width: number,
  height: number,
  color: { r: number; g: number; b: number; a?: number },
  bufferType: VideoBufferType = VideoBufferType.RGBA,
): VideoFrame {
  const channels = bufferType === VideoBufferType.RGB24 ? 3 : 4;
  const frameData = new Uint8Array(width * height * channels);

  for (let i = 0; i < frameData.length; i += channels) {
    frameData[i] = color.r;
    frameData[i + 1] = color.g;
    frameData[i + 2] = color.b;
    if (channels === 4 && color.a !== undefined) {
      frameData[i + 3] = color.a;
    }
  }

  return new VideoFrame(frameData, width, height, bufferType);
}

function createGradientFrame(width: number, height: number): VideoFrame {
  const channels = 4;
  const frameData = new Uint8Array(width * height * channels);

  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const idx = (y * width + x) * channels;
      frameData[idx] = Math.floor((x / (width - 1)) * 255);
      frameData[idx + 1] = Math.floor((y / (height - 1)) * 255);
      frameData[idx + 2] = 128;
      frameData[idx + 3] = 255;
    }
  }

  return new VideoFrame(frameData, width, height, VideoBufferType.RGBA);
}

function createPatternFrame(width: number, height: number, patterns: number[][]): VideoFrame {
  const channels = 4;
  const frameData = new Uint8Array(width * height * channels);

  for (let i = 0; i < patterns.length; i++) {
    const offset = i * 4;
    const pattern = patterns[i]!;
    frameData[offset] = pattern[0]!;
    frameData[offset + 1] = pattern[1]!;
    frameData[offset + 2] = pattern[2]!;
    frameData[offset + 3] = pattern[3]!;
  }

  return new VideoFrame(frameData, width, height, VideoBufferType.RGBA);
}

function verifyPngHeader(imageBuffer: Buffer) {
  expect(imageBuffer[0]).toBe(0x89);
  expect(imageBuffer[1]).toBe(0x50);
  expect(imageBuffer[2]).toBe(0x4e);
  expect(imageBuffer[3]).toBe(0x47);
}

function expectPixel(
  data: Buffer,
  index: number,
  expected: { r: number; g: number; b: number; a: number },
) {
  expect(data[index]).toBe(expected.r);
  expect(data[index + 1]).toBe(expected.g);
  expect(data[index + 2]).toBe(expected.b);
  expect(data[index + 3]).toBe(expected.a);
}

describe('computeChatCtxDiff', () => {
  it('should return empty operations for identical contexts', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');
    const msg2 = createChatMessage('2', 'Hi there', 'assistant');

    const oldCtx = createChatContext([msg1, msg2]);
    const newCtx = createChatContext([msg1, msg2]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    expect(result.toRemove).toEqual([]);
    expect(result.toCreate).toEqual([]);
  });

  it('should handle empty old context', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');
    const msg2 = createChatMessage('2', 'Hi there', 'assistant');

    const oldCtx = createChatContext([]);
    const newCtx = createChatContext([msg1, msg2]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    expect(result.toRemove).toEqual([]);
    expect(result.toCreate).toEqual([
      [null, '1'], // first item goes to root
      ['1', '2'],
    ]);
  });

  it('should handle empty new context', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');
    const msg2 = createChatMessage('2', 'Hi there', 'assistant');

    const oldCtx = createChatContext([msg1, msg2]);
    const newCtx = createChatContext([]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    expect(result.toRemove).toEqual(['1', '2']);
    expect(result.toCreate).toEqual([]);
  });

  it('should handle adding items to the end', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');
    const msg2 = createChatMessage('2', 'Hi there', 'assistant');
    const msg3 = createChatMessage('3', 'How are you?', 'user');

    const oldCtx = createChatContext([msg1, msg2]);
    const newCtx = createChatContext([msg1, msg2, msg3]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    expect(result.toRemove).toEqual([]);
    expect(result.toCreate).toEqual([['2', '3']]);
  });

  it('should handle removing items from the end', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');
    const msg2 = createChatMessage('2', 'Hi there', 'assistant');
    const msg3 = createChatMessage('3', 'How are you?', 'user');

    const oldCtx = createChatContext([msg1, msg2, msg3]);
    const newCtx = createChatContext([msg1, msg2]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    expect(result.toRemove).toEqual(['3']);
    expect(result.toCreate).toEqual([]);
  });

  it('should handle adding items to the beginning', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');
    const msg2 = createChatMessage('2', 'Hi there', 'assistant');
    const msg3 = createChatMessage('3', 'How are you?', 'user');

    const oldCtx = createChatContext([msg2, msg3]);
    const newCtx = createChatContext([msg1, msg2, msg3]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    expect(result.toRemove).toEqual([]);
    expect(result.toCreate).toEqual([[null, '1']]);
  });

  it('should handle removing items from the beginning', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');
    const msg2 = createChatMessage('2', 'Hi there', 'assistant');
    const msg3 = createChatMessage('3', 'How are you?', 'user');

    const oldCtx = createChatContext([msg1, msg2, msg3]);
    const newCtx = createChatContext([msg2, msg3]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    expect(result.toRemove).toEqual(['1']);
    expect(result.toCreate).toEqual([]);
  });

  it('should handle adding items in the middle', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');
    const msg2 = createChatMessage('2', 'Hi there', 'assistant');
    const msg3 = createChatMessage('3', 'How are you?', 'user');
    const msg4 = createChatMessage('4', 'Fine thanks', 'assistant');

    const oldCtx = createChatContext([msg1, msg3, msg4]);
    const newCtx = createChatContext([msg1, msg2, msg3, msg4]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    expect(result.toRemove).toEqual([]);
    expect(result.toCreate).toEqual([['1', '2']]);
  });

  it('should handle removing items from the middle', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');
    const msg2 = createChatMessage('2', 'Hi there', 'assistant');
    const msg3 = createChatMessage('3', 'How are you?', 'user');
    const msg4 = createChatMessage('4', 'Fine thanks', 'assistant');

    const oldCtx = createChatContext([msg1, msg2, msg3, msg4]);
    const newCtx = createChatContext([msg1, msg3, msg4]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    expect(result.toRemove).toEqual(['2']);
    expect(result.toCreate).toEqual([]);
  });

  it('should handle complex mixed operations', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');
    const msg2 = createChatMessage('2', 'Hi there', 'assistant');
    const msg3 = createChatMessage('3', 'How are you?', 'user');
    const msg4 = createChatMessage('4', 'Fine thanks', 'assistant');
    const msg5 = createChatMessage('5', 'Good to hear', 'user');
    const msg6 = createChatMessage('6', 'Anything else?', 'assistant');

    // Old: [1, 2, 3, 4]
    // New: [1, 5, 3, 6]
    // Remove: [2, 4]
    // Create: [5 after 1, 6 after 3]

    const oldCtx = createChatContext([msg1, msg2, msg3, msg4]);
    const newCtx = createChatContext([msg1, msg5, msg3, msg6]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    expect(result.toRemove).toEqual(['2', '4']);
    expect(result.toCreate).toEqual([
      ['1', '5'],
      ['3', '6'],
    ]);
  });

  it('should handle reordering items', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');
    const msg2 = createChatMessage('2', 'Hi there', 'assistant');
    const msg3 = createChatMessage('3', 'How are you?', 'user');

    // Old: [1, 2, 3]
    // New: [3, 1, 2]
    // This should remove all and recreate in new order

    const oldCtx = createChatContext([msg1, msg2, msg3]);
    const newCtx = createChatContext([msg3, msg1, msg2]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    // Since order changed completely, should have some operations
    expect(result.toRemove.length + result.toCreate.length).toBeGreaterThan(0);
  });

  it('should handle identical single item contexts', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');

    const oldCtx = createChatContext([msg1]);
    const newCtx = createChatContext([msg1]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    expect(result.toRemove).toEqual([]);
    expect(result.toCreate).toEqual([]);
  });

  it('should handle longest common subsequence correctly', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');
    const msg2 = createChatMessage('2', 'Hi there', 'assistant');
    const msg3 = createChatMessage('3', 'How are you?', 'user');
    const msg4 = createChatMessage('4', 'Fine thanks', 'assistant');
    const msg5 = createChatMessage('5', 'Good to hear', 'user');

    // Old: [1, 2, 3, 4, 5]
    // New: [1, 3, 5]
    // LCS should be [1, 3, 5], remove [2, 4]

    const oldCtx = createChatContext([msg1, msg2, msg3, msg4, msg5]);
    const newCtx = createChatContext([msg1, msg3, msg5]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    expect(result.toRemove).toEqual(['2', '4']);
    expect(result.toCreate).toEqual([]);
  });

  it('should handle interleaved additions and common subsequence', () => {
    const msg1 = createChatMessage('1', 'Hello', 'user');
    const msg2 = createChatMessage('2', 'Hi there', 'assistant');
    const msg3 = createChatMessage('3', 'How are you?', 'user');
    const msg4 = createChatMessage('4', 'Fine thanks', 'assistant');
    const msg5 = createChatMessage('5', 'Good to hear', 'user');
    const msg6 = createChatMessage('6', 'Anything else?', 'assistant');

    // Old: [1, 3, 5]
    // New: [1, 2, 3, 4, 5, 6]
    // LCS: [1, 3, 5], add [2 after 1, 4 after 3, 6 after 5]

    const oldCtx = createChatContext([msg1, msg3, msg5]);
    const newCtx = createChatContext([msg1, msg2, msg3, msg4, msg5, msg6]);

    const result = computeChatCtxDiff(oldCtx, newCtx);

    expect(result.toRemove).toEqual([]);
    expect(result.toCreate).toEqual([
      ['1', '2'],
      ['3', '4'],
      ['5', '6'],
    ]);
  });
});

describe('serializeImage', () => {
  let consoleWarnSpy: ReturnType<typeof vi.spyOn>;

  beforeEach(() => {
    consoleWarnSpy = vi.spyOn(console, 'warn').mockImplementation(() => {});
    vi.clearAllMocks();
  });

  describe('Data URL handling', () => {
    it('should serialize a valid JPEG data URL', async () => {
      const originalBase64 = '/9j/4AAQSkZJRg==';
      const imageContent = createImageContent(`data:image/jpeg;base64,${originalBase64}`, 'high');

      const result = await serializeImage(imageContent);

      expect(result).toEqual({
        base64Data: originalBase64,
        mimeType: 'image/jpeg',
        inferenceDetail: 'high',
      });

      expect(result.base64Data).toBe(originalBase64);
    });

    it('should serialize a valid PNG data URL', async () => {
      const imageContent = createImageContent(
        'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAAB',
        'low',
      );

      const result = await serializeImage(imageContent);

      expect(result).toEqual({
        base64Data: 'iVBORw0KGgoAAAANSUhEUgAAAAEAAAAB',
        mimeType: 'image/png',
        inferenceDetail: 'low',
      });
    });

    it('should serialize a valid WebP data URL', async () => {
      const imageContent = createImageContent(
        'data:image/webp;base64,UklGRiQAAABXRUJQVlA4IBgAAAA',
        'auto',
      );

      const result = await serializeImage(imageContent);

      expect(result).toEqual({
        base64Data: 'UklGRiQAAABXRUJQVlA4IBgAAAA',
        mimeType: 'image/webp',
        inferenceDetail: 'auto',
      });
    });

    it('should serialize a valid GIF data URL', async () => {
      const imageContent = createImageContent(
        'data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==',
        'auto',
      );

      const result = await serializeImage(imageContent);

      expect(result).toEqual({
        base64Data: 'R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==',
        mimeType: 'image/gif',
        inferenceDetail: 'auto',
      });
    });

    it('should warn and use provided mimeType when it differs from data URL mime type', async () => {
      const imageContent = createImageContent('data:image/jpeg;base64,/9j/4AAQSkZJRg==', 'auto', {
        mimeType: 'image/png',
      });

      const result = await serializeImage(imageContent);

      expect(consoleWarnSpy).toHaveBeenCalledWith(
        "Provided mimeType 'image/png' does not match data URL mime type 'image/jpeg'. Using provided mimeType.",
      );
      expect(result).toEqual({
        base64Data: '/9j/4AAQSkZJRg==',
        mimeType: 'image/png',
        inferenceDetail: 'auto',
      });
    });

    it('should throw error for invalid data URL format', async () => {
      const imageContent = createImageContent('data:;base64,/9j/4AAQSkZJRg==', 'auto');

      await expect(serializeImage(imageContent)).rejects.toThrow('Invalid data URL format');
    });

    it('should throw error for unsupported mime type', async () => {
      const imageContent = createImageContent(
        'data:image/bmp;base64,Qk06AAAAAAAAADYAAAAoAAAA',
        'auto',
      );

      await expect(serializeImage(imageContent)).rejects.toThrow(
        'Unsupported mimeType image/bmp. Must be jpeg, png, webp, or gif',
      );
    });
  });

  describe('External URL handling', () => {
    it('should serialize an external URL without mimeType', async () => {
      const imageContent = createImageContent('https://example.com/image.jpg', 'high');

      const result = await serializeImage(imageContent);

      expect(result).toEqual({
        mimeType: undefined,
        inferenceDetail: 'high',
        externalUrl: 'https://example.com/image.jpg',
      });
    });

    it('should serialize an external URL with mimeType', async () => {
      const imageContent = createImageContent('https://example.com/image.jpg', 'low', {
        mimeType: 'image/jpeg',
      });

      const result = await serializeImage(imageContent);

      expect(result).toEqual({
        mimeType: 'image/jpeg',
        inferenceDetail: 'low',
        externalUrl: 'https://example.com/image.jpg',
      });
    });
  });

  describe('VideoFrame handling', () => {
    it('should serialize a VideoFrame without resize parameters', async () => {
      const width = 4;
      const height = 4;
      const videoFrame = createSolidColorFrame(width, height, { r: 255, g: 0, b: 0, a: 255 });
      const imageContent = createImageContent(videoFrame, 'auto');

      const result = await serializeImage(imageContent);

      expect(result).toMatchObject({
        mimeType: 'image/png',
        inferenceDetail: 'auto',
      });
      expect(result.base64Data).toBeDefined();
      expect(result.base64Data).toMatch(/^[A-Za-z0-9+/]+=*$/);
      expect(result.externalUrl).toBeUndefined();

      const { imageBuffer, decodedImage } = await decodeImageToRaw(result.base64Data!);

      verifyPngHeader(imageBuffer);

      expect(decodedImage.info.width).toBe(width);
      expect(decodedImage.info.height).toBe(height);
      expect(decodedImage.info.channels).toBe(4);

      const decodedData = decodedImage.data;
      for (let i = 0; i < decodedData.length; i += 4) {
        expectPixel(decodedData, i, { r: 255, g: 0, b: 0, a: 255 });
      }
    });

    it('should serialize a VideoFrame with a gradient pattern', async () => {
      const width = 8;
      const height = 8;
      const videoFrame = createGradientFrame(width, height);
      const imageContent = createImageContent(videoFrame, 'high');

      const result = await serializeImage(imageContent);

      expect(result).toMatchObject({
        mimeType: 'image/png',
        inferenceDetail: 'high',
      });

      const { decodedImage } = await decodeImageToRaw(result.base64Data!);

      expect(decodedImage.info.width).toBe(width);
      expect(decodedImage.info.height).toBe(height);

      const decodedData = decodedImage.data;

      for (let y = 0; y < height; y++) {
        for (let x = 0; x < width; x++) {
          const idx = (y * width + x) * 4;
          const expectedR = Math.floor((x / (width - 1)) * 255);
          const expectedG = Math.floor((y / (height - 1)) * 255);

          expectPixel(decodedData, idx, { r: expectedR, g: expectedG, b: 128, a: 255 });
        }
      }
    });

    it('should preserve exact pixel data through serialization', async () => {
      const width = 2;
      const height = 2;

      const patterns = [
        [255, 0, 0, 255],
        [0, 255, 0, 255],
        [0, 0, 255, 255],
        [255, 255, 255, 255],
      ];

      const videoFrame = createPatternFrame(width, height, patterns);
      const imageContent = createImageContent(videoFrame, 'low');

      const result = await serializeImage(imageContent);

      const { decodedImage } = await decodeImageToRaw(result.base64Data!);
      const decoded = decodedImage.data;

      expectPixel(decoded, 0, { r: 255, g: 0, b: 0, a: 255 });
      expectPixel(decoded, 4, { r: 0, g: 255, b: 0, a: 255 });
      expectPixel(decoded, width * 4, { r: 0, g: 0, b: 255, a: 255 });
      expectPixel(decoded, (width + 1) * 4, { r: 255, g: 255, b: 255, a: 255 });
    });

    it('should handle resize parameters correctly', async () => {
      const width = 2;
      const height = 2;
      const videoFrame = createSolidColorFrame(width, height, { r: 100, g: 100, b: 100, a: 255 });

      const imageContent = createImageContent(videoFrame, 'auto', {
        inferenceWidth: 4,
        inferenceHeight: 4,
      });

      const result = await serializeImage(imageContent);

      const { decodedImage } = await decodeImageToRaw(result.base64Data!);

      expect(decodedImage.info.width).toBe(4);
      expect(decodedImage.info.height).toBe(4);

      const decodedData = decodedImage.data;
      for (let i = 0; i < decodedData.length; i += 4) {
        expect(decodedData[i]).toBeCloseTo(100, -1);
        expect(decodedData[i + 1]).toBeCloseTo(100, -1);
        expect(decodedData[i + 2]).toBeCloseTo(100, -1);
        expect(decodedData[i + 3]).toBe(255);
      }
    });

    it('should handle RGB24 VideoBufferType correctly', async () => {
      const width = 2;
      const height = 2;
      const channels = 3;
      const frameData = new Uint8Array(width * height * channels);

      for (let i = 0; i < frameData.length; i += channels) {
        frameData[i] = 255;
        frameData[i + 1] = 128;
        frameData[i + 2] = 64;
      }

      const videoFrame = new VideoFrame(frameData, width, height, VideoBufferType.RGB24);
      const imageContent = createImageContent(videoFrame, 'auto');

      const result = await serializeImage(imageContent);

      expect(result.mimeType).toBe('image/png');

      const { decodedImage } = await decodeImageToRaw(result.base64Data!);

      expect(decodedImage.info.channels).toBeGreaterThanOrEqual(3);

      const decodedData = decodedImage.data;
      const decodedChannels = decodedImage.info.channels;

      for (let i = 0; i < decodedData.length; i += decodedChannels) {
        expect(decodedData[i]).toBe(255);
        expect(decodedData[i + 1]).toBe(128);
        expect(decodedData[i + 2]).toBe(64);
        if (decodedChannels === 4) {
          expect(decodedData[i + 3]).toBe(255);
        }
      }
    });

    it('should handle different RGBA-like formats correctly', async () => {
      const width = 1;
      const height = 1;
      const testFormats = [
        VideoBufferType.RGBA,
        VideoBufferType.BGRA,
        VideoBufferType.ARGB,
        VideoBufferType.ABGR,
      ];

      for (const format of testFormats) {
        const frameData = new Uint8Array([100, 150, 200, 250]);
        const videoFrame = new VideoFrame(frameData, width, height, format);
        const imageContent = createImageContent(videoFrame, 'auto');

        const result = await serializeImage(imageContent);

        expect(result.mimeType).toBe('image/png');
        expect(result.base64Data).toBeDefined();
      }
    });
  });

  describe('Error handling', () => {
    it('should throw error for unsupported image type', async () => {
      const imageContent = createImageContent(123 as any, 'auto'); // eslint-disable-line @typescript-eslint/no-explicit-any

      await expect(serializeImage(imageContent)).rejects.toThrow('Unsupported image type');
    });
  });
});
