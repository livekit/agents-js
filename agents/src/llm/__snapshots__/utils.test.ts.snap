// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`oaiParams > common types 1`] = `
{
  "additionalProperties": false,
  "properties": {
    "address": {
      "additionalProperties": false,
      "properties": {
        "city": {
          "type": "string",
        },
        "state": {
          "type": "string",
        },
        "street": {
          "type": "string",
        },
        "zip": {
          "type": "string",
        },
      },
      "required": [
        "street",
        "city",
        "state",
        "zip",
      ],
      "type": "object",
    },
    "age": {
      "type": "number",
    },
    "favoriteColor": {
      "enum": [
        "red",
        "green",
        "blue",
      ],
      "type": "string",
    },
    "hobbies": {
      "items": {
        "type": "string",
      },
      "type": "array",
    },
    "isStudent": {
      "type": "boolean",
    },
    "name": {
      "type": "string",
    },
  },
  "required": [
    "name",
    "age",
    "isStudent",
    "hobbies",
    "favoriteColor",
    "address",
  ],
  "type": "object",
}
`;
